### PR TITLE
[LoopSchedule] Add initial LoopSchedule dialect

### PR DESCRIPTION
**Pull request by andrewb1999**
This is the first PR in a series of PRs to add the LoopSchedule dialect as discussed in a recent CIRCT ODM. This dialect will be the new home for the `pipeline.while` op, as well as new ops to represent scheduled sequential loops. The RFC can be found [here](https://discourse.llvm.org/t/rfc-split-pipeline-dialect-and-add-representation-for-sequential-loop-scheduling/69294).

### CODE REVIEW

File: docs/Dialects/LoopSchedule/LoopSchedule.md [hunk 1]
```diff
@@ -0,0 +1,15 @@
+# LoopSchedule Dialect Rationale
+
+This document describes various design points of the `loopschedule` dialect, why it is
+the way it is, and current status. This follows in the spirit of other [MLIR
+Rationale docs](https://mlir.llvm.org/docs/Rationale/).
+
+## Introduction
+
+The `loopschedule` dialect provides a collection of ops to represent software-like loops after scheduling. There are currently two main kinds of loops that can be represented, pipelined and sequential. Pipelined loops allow multiple iterations of the loop to be in-flight at a time and have an associated initiation interval (II) to specify the number of cycles between the start of successive loop iterations. In contrast, sequential loops are guaranteed to only have one iteration in-flight at any given time.
```
Comment (mikeurbach) [inline docs/Dialects/LoopSchedule/LoopSchedule.md:9]: nit: If you're going to wrap the first paragraph, I think you should wrap every paragraph.

Comment (mikeurbach) [reply docs/Dialects/LoopSchedule/LoopSchedule.md:9]: also nit: `represented, pipelined and sequential` -> `represented: pipelined and sequential`

File: lib/Dialect/LoopSchedule/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+add_circt_dialect_library(CIRCTLoopSchedule
+  LoopScheduleOps.cpp
+
+  ADDITIONAL_HEADER_DIRS
+  ${PROJECT_SOURCE_DIR}/include
+  ${PROJECT_BINARY_DIR}/include
+
+  LINK_LIBS PUBLIC
+  CIRCTESI
```
Comment (mikeurbach) [inline lib/Dialect/LoopSchedule/CMakeLists.txt:15]: Does this dialect need a dependency on ESI?

**Review by mikeurbach — APPROVED** [review]
The dialect boilerplate looks good to me, just a few minor comments.

File: lib/Dialect/LoopSchedule/CMakeLists.txt (hunk 2 above)
Comment (andrewb1999) [reply lib/Dialect/LoopSchedule/CMakeLists.txt:15]: No, this is a mistake. I will fix it.

**Comment by mikeurbach** [comment]
Thanks for the updates, please feel free to merge once CI passes.

**Review by mortbopet — APPROVED** [review]

**Pull request merged**